### PR TITLE
Add metrics (writes and error counts, timing) for consumer mode

### LINKED ISSUE #36: Add observability to consume mode
The consume mode currently is not exposing any status of how successful its remote_write requests.

Ideally it should expose metrics about the outcome of the remote_write requests in a similar way as the "produce" mode does.

The pulsar client itself registers some metrics about consume the messages from the bus.

Their metrics definition is here: https://github.com/grafana/prometheus-pulsar-remote-write/blob/ed7be597220fce8293d287856807551ec9a93803/pkg/app/metrics.go#L8

Ideally the metric vector includes the tenant ID if set, so we can figure which tenant fails to deliver remote_write requests

### PR DESCRIPTION
**Pull request by 56quarters**
Fixes #36

### CODE REVIEW

File: pkg/metrics/consume.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type ConsumerMetrics struct {
+	WrittenSamples            prometheus.Counter
```
Comment (simonswine) [inline pkg/metrics/consume.go:9]: I think it probably make sense to have this metric as a Vector with a per tenant value as well

File: pkg/metrics/consume.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type ConsumerMetrics struct {
```
Comment (simonswine) [inline pkg/metrics/consume.go:8]: I am not too sure about the naming of having `Sent` and `Written` as two separate metrics. I think we could potentially use the same `*Metric struct` for both consumer and producer. The value of the tenant label could e.g. just stay empty as could the remote for the producer.

File: pkg/app/consume.go [hunk 3]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"github.com/grafana/prometheus-pulsar-remote-write/pkg/metrics"
```
Comment (simonswine) [inline pkg/app/consume.go:7]: nit: Seems like the golangci-lint is not enabled for checking on that, but usually we try to follow the import ordering as described here:  https://github.com/uber-go/guide/blob/master/style.md#import-group-ordering

**Review by simonswine — COMMENTED** [review]
Nice one @56quarters that looks already pretty close. See my comments in line.

I think it would be great if we could cover the metrics as part of the integration test through https://github.com/prometheus/client_golang/tree/master/prometheus/testutil

File: pkg/metrics/consume.go (hunk 1 above)
Comment (56quarters) [reply pkg/metrics/consume.go:9]: Will do

File: pkg/app/consume.go (hunk 3 above)
Comment (56quarters) [reply pkg/app/consume.go:7]: Will change

File: pkg/metrics/consume.go (hunk 2 above)
Comment (56quarters) [reply pkg/metrics/consume.go:8]: OK, I can combine them. Do you think it makes sense to have a single struct for both producer and consumer and then have each command only operate on a subset of the metrics?

File: pkg/metrics/consume.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	WrittenSamples            prometheus.Counter
+	RecoverableWriteErrors    *prometheus.CounterVec
+	NonRecoverableWriteErrors *prometheus.CounterVec
+	WriteBatchDuration        *prometheus.HistogramVec
+}
+
+func NewConsumerMetricsDefault() *ConsumerMetrics {
```
[the remark this answers is not in the document]
Comment (56quarters) [reply pkg/metrics/consume.go:15]: Right, I will change to to `Nop` to match the default logger used by the `Write` command if I keep them. If not using these constructors, what would be your preference? A single constructor function that callers could pass `nil` into?

File: pkg/metrics/consume.go (hunk 2 above)
Comment (56quarters) [reply pkg/metrics/consume.go:8]: Or use the existing metrics? On second glance, they do seem pretty general and could make sense for producer or consumer mode.

**Comment by 56quarters** [comment]
I'll take a look at testing metrics as part of the integration tests.

File: pkg/metrics/consume.go (hunk 2 above)
Comment (56quarters) [reply pkg/metrics/consume.go]: I've taken the approach of using the same metrics for both producer and consumer in the latest commit. I added a retryable error metric as well.

Comment (simonswine) [reply pkg/metrics/consume.go]: That looks great :+1:  and we also get the tenant IDs now in the produce part

File: pkg/app/app.go [hunk 5]
```diff
@@ -65,11 +68,13 @@ func New() *App {
 	app.Flag("web.max-connection-age", "If set this limits the maximum lifetime of persistent HTTP connections.").
 		Default("0s").DurationVar(&cfg.maxConnectionAge)
 
+	reg := prometheus.NewRegistry()
```
Comment (simonswine) [inline pkg/app/app.go:71]: Creating a new registry, prevents showing the standard golang metrics (of memory, cpu, goroutines) and also doesn't show the metrics of the pulsar library which uses the DefaultRegisterer. 

Ideally we could register the standard metrics (`NewGoCollector` and `NewProcessCollector`) and pass the registry to pulsar (I haven't checked if pulsar supports it)

File: integration/consume_integration_test.go [hunk 6]
```diff
@@ -225,6 +230,19 @@ func (ti *testConsumeIntegration) test(t *testing.T) {
...
+	// there shouldn't be any errors).
+	actualMetrics, err := testutil.GatherAndCount(
+		ti.app.Gatherer(),
+		"received_samples_total",
+		"sent_samples_total",
+		"sent_batch_duration_seconds",
+	)
+
+	assert.NoError(t, err)
+	assert.Equal(t, ti.expectedMetrics, actualMetrics, "expected "+strconv.Itoa(ti.expectedMetrics)+" to be updated during consume test")
```
Comment (simonswine) [inline integration/consume_integration_test.go:244]: nit: I personally find `fmt.Sprintf` more readable than `strconv`

File: integration/produce_integration_test.go [hunk 7]
```diff
@@ -191,17 +197,34 @@ func (ti *testProduceIntegration) test(t *testing.T) {
...
+	// there shouldn't be any errors).
+	actualMetrics, err := testutil.GatherAndCount(
+		ti.app.Gatherer(),
+		"received_samples_total",
+		"sent_samples_total",
+		"sent_batch_duration_seconds",
+	)
+
+	assert.NoError(t, err)
+	assert.Equal(t, ti.expectedMetrics, actualMetrics, "expected "+strconv.Itoa(ti.expectedMetrics)+" metrics to be updated during produce test")
```
Comment (simonswine) [inline integration/produce_integration_test.go:211]: nit: I personally find `fmt.Sprintf` more readable than `strconv`

**Review by simonswine — COMMENTED** [review]
I think it's almost there. See my comments relating the `GoCollector` and `ProcessCollector` metrics

File: integration/produce_integration_test.go (hunk 7 above)
Comment (56quarters) [reply integration/produce_integration_test.go:211]: Will change.

File: integration/consume_integration_test.go (hunk 6 above)
Comment (56quarters) [reply integration/consume_integration_test.go:244]: Will change

File: pkg/app/app.go (hunk 5 above)
Comment (56quarters) [reply pkg/app/app.go:71]: I'll look into this.

File: integration/consume_integration_test.go [hunk 8]
```diff
@@ -225,6 +230,19 @@ func (ti *testConsumeIntegration) test(t *testing.T) {
 		}
 	}
 
+	// Check the number of metrics incremented during this test (for consume mode, we expect
+	// three metrics with a single value for each label: two counts and the timing histogram since
+	// there shouldn't be any errors).
+	actualMetrics, err := testutil.GatherAndCount(
```
Comment (simonswine) [inline integration/consume_integration_test.go:236]: I might maybe also provide some value, if we lint the metrics a bit like that:

```diff
+       lintProblems, err := testutil.GatherAndLint(
+               ti.app.Gatherer(),
+               "received_samples_total",
+               "sent_samples_total",
+               "sent_batch_duration_seconds",
+       )
```

Comment (56quarters) [reply integration/consume_integration_test.go:236]: Will add

**Review by simonswine — APPROVED** [review]
LGTM, thanks for the metrics :+1:

**Pull request merged**